### PR TITLE
Add somewhat working primary virtio queue support.

### PR DESCRIPTION
**Pull request by dmgolubovsky**
Signed-off-by: golubovsky <golubovsky@gmail.com>

This PR adds primary support functions for the virtqueue transport. Common functionality provides the namespace for virtqueues as files that can be opened for sending and receiving data. Each type of a virtual device needs a specific "flavor" functions to be written (for reading, writing, and cleaning up device-specific data at the moment; specialized initialization functionality can be embedded in the flavor's "constructor" function). Flavor for 9p is provided, but per the recent discussion (bug #324) it is unlikely to work properly with the present QEMU implementation of 9p. Also provided are basic functions to borrow virtio buffer descriptors from the queue, placing them in the queue, and handling interrupts. Testing was limited to the attempts to implement 9p.

### CODE REVIEW

File: sys/src/9/amd64/devvq.c [hunk 1]
```diff
@@ -0,0 +1,638 @@
...
+// Map device identifiers to descriptive strings to display as virtio device names.
+
+typedef struct
+{
+	uint16_t did;
+	char *desc;
+	Flavor *flv;
+} didmap;
+
+static didmap dmtab[] = {
```
Comment (rminnich) [inline sys/src/9/amd64/devvq.c]: The virtio spec defines a basic transport. I think there's a bit too many things in this file. The plan 9 way to implement this type of model would be to define the basic virtio transport, and then define the types of transports that use it as different entities. Consider the way that tcp/ip stack works. There are ethernet devices, and there are network stacks. To get a stack to use a device, you do something like
echo ether0 > /net/ipifc/0/ctl
I think there are too many levels squished in here. I think that is indicated by the fact that you need this dmtab -- it just doesn't feel right. Could you maybe just implement the the basic virtio queue abstraction and then we can look at implementing the various tranports in other drivers? That's more the style of plan 9. 
Or, maybe, look at the storage device abstraction (devsd) and how the different types of drivers (sdata, sdscsi, etc.) link into it. This is really great progress but the presence of this dmtab just doesn't seem right.

or just consider the relationship of #$ (pci device) to PCI devices. #$ is a generic driver for managing the pci abstraction, and individual devices might use it.

File: sys/src/9/386/pci.c [hunk 2]
```diff
@@ -193,7 +193,6 @@ pcilscan(int bno, char *path, Pcidev** list)
 				}
 				break;
 
-			case 0x00:
```
Comment (dmgolubovsky) [inline sys/src/9/386/pci.c]: This was part of PR #180 (Some virtio-pci devices (e. g. 9p)  have ccrb = 0x00, their BARs and sizes also shouldbe  picked up here) which was reverted (probably by omission). It has to do with virtio of course.

File: sys/src/9/amd64/vq9p.c [hunk 3]
```diff
@@ -0,0 +1,216 @@
+/*
```
Comment (dmgolubovsky) [inline sys/src/9/amd64/vq9p.c]: It is probably intel-specific as is all virtio (does Linux on ARM provide identical functionality in terms of virtualization, and is it in the same wide use as on amd64)?

File: sys/src/9/amd64/devvq.c [hunk 4]
```diff
@@ -0,0 +1,638 @@
...
+
+#include	"virtio_custom.h"
+
+#define TYPE(q)			((uint32_t)(q).path & 0x0F)
+#define DEV(q)			((uint32_t)(((q).path >> 4) & 0x0FFF))
+#define VQ(q)			((uint32_t)(((q).path >> 16) & 0x0FFFF))
+#define QID(c, t)		((((c) & 0x0FFF)<<4) | ((t) & 0x0F))
+#define VQQID(q, c, t)	((((q) & 0x0FFFF)<<16) | (((c) & 0x0FFF)<<4) | ((t) & 0x0F))
+
+#define BY2PG PGSZ
```
Comment (dmgolubovsky) [inline sys/src/9/amd64/devvq.c]: is this about BY2PG? I think it is one remnant from 9front code which used this constant, can be changed.

File: sys/src/9/amd64/devvq.c [hunk 5]
```diff
@@ -0,0 +1,638 @@
...
+	Qvqdata,				// virtqueue file entry for reading-writing messages
+	QVqctl,					// virtqueue file entry for control operations, write-only
+};
+
+enum
+{
+	Vqqsize = 64*1024		// default size for the device read queue size
+};
+
+static Dirtab topdir[] = {
```
Comment (dmgolubovsky) [inline sys/src/9/amd64/devvq.c]: I would not unify them with #| simply because as we see in the 9p example they may be far from dumb pipes. Virtio-net has command and data queues, and they make no sense to be used separately. Same with virtio-sd, they are specialized. Some virtio devices may open their queues for direct read-write, some not. My current code (via per-device flavor) allows for unified device identification and discovery, but if the flavor implementation blocks read/write access to virtqueues as simply files (pipes) then they will not be accessible from regular processes (though visible as device metadata).

Comment (dmgolubovsky) [inline sys/src/9/amd64/devvq.c]: Or maybe I misunderstand the question about #|, can you clarify?

File: sys/src/9/amd64/vq9p.c (hunk 3 above)
Comment (rminnich) [inline sys/src/9/amd64/vq9p.c]: yeah. We've just finished a port of virtio to akaros and there's almost nothing architecture dependent in it. Is there anything in your code that is x86 specific? I don't see anything.

File: sys/src/9/amd64/devvq.c [hunk 6]
```diff
@@ -0,0 +1,638 @@
+/*
```
Comment (rminnich) [inline sys/src/9/amd64/devvq.c]: I don't see anything about virtio that is architecture-specific. Note that rusty russell had a lot to do with virtio and at that time he was an IBM Power guy.

File: sys/src/9/amd64/devvq.c (hunk 4 above)
Comment (rminnich) [inline sys/src/9/amd64/devvq.c]: it needs to be changed.

File: sys/src/9/amd64/devvq.c (hunk 5 above)
Comment (rminnich) [inline sys/src/9/amd64/devvq.c]: you misunderstood the question. 
#| is a simple bidirectional pair of pipes. It's a nice abstraction. 

If we ignore implementation details, virtio is a similar abstraction. I think you ought to be able to define a basic abstraction, which is virtio, that is similar in nature to pipes, in that its function is to provide a way to move data. 

What i'm trying to say is that I'd like to see this PR implement the basic virtio abstraction, i.e. moving data, and then we can add the flavors in a later PR. I don't think that should all be merged in the code and in this PR.

File: sys/src/9/amd64/devvq.c (hunk 1 above)
Comment (rminnich) [inline sys/src/9/amd64/devvq.c]: I think almost all of this code can be kept. If you want I can make some simple changes to implement what I'm talking about.

Comment (rminnich) [inline sys/src/9/amd64/devvq.c]: OK, think about it this way. The pipe device provides two bidirectional queues. Why can't the virtio device, similarly, provide a single bidirectional queue? That's a basic abstraction -- a queue. That's how it is intended to work per the spec. You can then build on that abstraction and implement other things, such as net, 9p, console, disk, etc. 

So how about we start with something that implements just the queue?

Comment (rminnich) [inline sys/src/9/amd64/devvq.c]: Also, your vqlib idea is pretty clever.

File: sys/src/9/amd64/vq9p.c (hunk 3 above)
Comment (dmgolubovsky) [inline sys/src/9/amd64/vq9p.c]: Not x86 specific in the code, but my understanding was that virtio is mainly available and usable on x86 platforms (s390 which is also mentiond in the spec left alone). Plus, it relies on pci, and pci.c is in 386, though devpci.c is in port. That made me think about intel-specificity of virtio-pci. I just cannot see anything like pci.c in aarch64...

File: sys/src/9/386/pci.c (hunk 2 above)
Comment (dmgolubovsky) [inline sys/src/9/386/pci.c]: Perhaps I need to send a separate PR on this?

File: sys/src/9/amd64/devvq.c (hunk 1 above)
Comment (dmgolubovsky) [inline sys/src/9/amd64/devvq.c]: OK, my research in the implementation of virtio (esp. how 9p, sd, and net are implemented - they do not expose any "simple" queues per se) made me think that this is somewhat different from just a pipe (although very similar). I'll try to study the logic of #| and what was done in Akaros. I'll push the code for vqlib (these few functions as much as I can separate them) to place them in port, just need to understand better what handles PCI for ARM in Harvey.

Thanks.

**Comment by rminnich** [comment]
My point here is we don't need to limit our thinking. The virtqueue is a nice construct, as it is designed to work in a very efficient manner. The virtqueue, considered alone, could be useful for many things. We should not limit our idea about possible uses to what the linux/kvm world is using it for. Also, the way things are done in plan 9 is not at all the way they're done in the unix world, so we need to integrate this new construct into harvey in a way that makes sense for harvey. 

I don't think this requires lots of changes in what you have here; most of what I'm advocating is a removal of some of the code, but most of this PR looks really good.

as for PCI, the reasons that some of pci is in 386/ is that the 0xcf8/0xcfc configuration space mechanism is very x86 specific. But devpci can be in port/ because a lot of pci is NOT x86 specific. Similarly, I still feel that all the code you have written can go in port/ instead of amd64/.

**Comment by dmgolubovsky** [comment]
This PR is not going to proceed, so closing it now.
